docs($http): expand caching section to cover FAQ

ryanhart2 opened the pull request:
- add note that only GET & JSONP requests are cached
- add note that cache keys are the url and search params (headers not considered)
- add note that cache-control header on response does not affect angular caching
- add note that $httpProvider.defaults.cache can be used to set cache default (in addition to $http.defaults.cache)
- expand note on how default cache property affects requests with cache set to true

Fixes #11101

gkalpak commented on src/ng/http.js:
```diff
@@ -608,26 +608,30 @@ function $HttpProvider() {
+     *
+     *   * Only GET and JSONP requests are cached.
+     *   * The cache key is the request url including search parameters; headers are not considered.
+     *   * Cached responses are return asynchronously, in the same way as responses from the server.
```
_return_  -->  _returned_

ryanhart2 commented:
@gkalpak, I fixed the spelling error and also uppercased 'URL' to be consistent with the rest of the comments. I also added Fixes #11101 to the commit comment.

gkalpak commented on src/ng/http.js:
```diff
@@ -608,26 +608,30 @@ function $HttpProvider() {
      *
-     * If you set the default cache to `false` then only requests that specify their own custom
-     * cache object will be cached.
+     * The default cache can also be set to `false`, which will mean that only requests that specify
```
This sentence seems incorrect (from taking a quick look at the code).

ryanhart2 commented:
@gkalpak, you are correct: a default cache value of false can be overridden by a cache config value of true as well as a custom cache object, so I will update.

This means that rows 260-263 of the file (describing the defaults object) are also incorrect, so I will include a revision of these in my commit as well.

There is also no mention of the effect of setting the default cache to true (it results in the $http cache being used when the config cache is undefined), so I will expand my comments to include this.

I set up this [plunker] (http://plnkr.co/edit/QuHeEzUYLmyKNxuIjMUh?p=preview) to test the various combinations.

Narretz commented:
@ryanhart2 Did you ever update this PR?

ryanhart2 commented:
@Narretz Apologies, I didn't and had forgotten about it. I will have a look this weekend and try to get it done.

Narretz commented:
Am 21.01.2016 um 15:04 schrieb Ryan Hart:
>
> @Narretz <https://github.com/Narretz> Apologies, I didn't and had 
> forgotten about it. I will have a look this weekend and try to get it 
> done.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/angular/angular.js/pull/13003#issuecomment-173577789>.
>
No problem. I just wouldn't want to see good work go to waste!

Narretz commented:
@ryanhart2 why did you close it?

Commit:
add note that only GET & JSONP requests are cached
add note that cache keys are the url and search params (headers not considered)
add note that cache-control header on response does not affect angular caching
add note that $httpProvider.defaults.cache can be used to set cache default (in addition to $http.defaults.cache)
revise explanation of how caching can be enabled using cache.config value and cache default value
Fixes #11101

ryanhart2 commented:
@Narretz Just struggling to do the rebase... I think I am getting there!

ryanhart2 commented:
@Narretz , @gkalpak I have updated this pull request to more accurately describe how http caching works. Appreciate if you have time to review and provide feedback. Thanks.

gkalpak commented on src/ng/http.js:
```diff
@@ -805,10 +812,8 @@ function $HttpProvider() {
-     *      GET request, otherwise if a cache instance built with
-     *      {@link ng.$cacheFactory $cacheFactory}, this cache will be used for
-     *      caching.
+     *    - **cache** – `{boolean|Object}` – A value to enable caching of the HTTP response. See
```
IMO it is still useful to mention that the value must be either boolean or a `Cache` object created with `$cacheFactory`.

gkalpak commented on src/ng/http.js:
```diff
@@ -257,10 +257,8 @@ function $HttpProvider() {
-   * that will provide the cache for all requests who set their `cache` property to `true`.
-   * If you set the `defaults.cache = false` then only requests that specify their own custom
-   * cache object will be cached. See {@link $http#caching $http Caching} for more information.
+   * - **`defaults.cache`** - {boolean|Object} - A value to enable caching of HTTP responses
```
IMO it is still useful to mention that the value must be either boolean or a `Cache` object created with `$cacheFactory`.

gkalpak commented:
I left a couple of comments (actually it's the same comment twice :stuck_out_tongue: ).
BTW, @ryanhart2, you don't need to rebase. (In fact it's preferrable to just push a new commit with the changes :smiley:)

Commit:
Update pull request to mention that defaults.cache or config.cache value must be either boolean or a Cache object created with $cacheFactory

ryanhart2 commented:
Thanks @gkalpak. I agree, the revised version now reads better. Thanks for the hint on the rebase. I have pushed a new commit addressing your suggestions. Note that, on the same lines, I also updated the words "enable caching" to "enable or disable caching", given that the values could be `false`.

gkalpak commented:
Backported to `v1.5.x` (2a7d4c1) and `v1.4.x` (b9d3625).
Thx @ryanhart2 for working on this !
